ooniprobe/Utility/TestUtility.mm

```diff
@@ -86,24 +86,31 @@ + (void)downloadUrls:(void (^)(NSArray *))completion {
         path = [NSString stringWithFormat:@"%@&category_codes=%@", path, [categories componentsJoinedByString:@","]];
     }
     NSURL *url = [NSURL URLWithString:path];
-    NSURLSessionDataTask *downloadTask = [[NSURLSession sharedSession] dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-        if (!error) {
-            NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
-            NSArray *urlsArray = [dic objectForKey:@"results"];
-            NSMutableArray *urls = [[NSMutableArray alloc] init];
-            for (NSDictionary* current in urlsArray){
-                //List for database
-                Url *url = [Url checkExistingUrl:[current objectForKey:@"url"] categoryCode:[current objectForKey:@"category_code"] countryCode:[current objectForKey:@"country_code"]];
-                //List for mk
-                [urls addObject:url.url];
-            }
-            completion(urls);
-        }
-        else {
-            // Fail
-            completion(nil);
-            NSLog(@"error : %@", error.description);
-        }
+    NSURLSessionDataTask *downloadTask = [[NSURLSession sharedSession]
+     dataTaskWithURL:url
+     completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+         if (error != nil) {
+             errorcb(error);
+             return;
+         }
+         NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
+         if (error != nil) {
+             errorcb(error);
+             return;
+         }
+         NSArray *urlsArray = [dic objectForKey:@"results"];
+         NSMutableArray *urls = [[NSMutableArray alloc] init];
+         for (NSDictionary* current in urlsArray){
+             //List for database
+             Url *url = [Url checkExistingUrl:[current objectForKey:@"url"] categoryCode:[current objectForKey:@"category_code"] countryCode:[current objectForKey:@"country_code"]];
+             //List for mk
+             [urls addObject:url.url];
+         }
+         if ([urls count] == 0){
+             errorcb(error);
+             return;
+         }
```

Comment:
This check should be moved before looping over the array.

Reply:
It's correct like it is, review the code, `urls` is an array I initialize in the loop

Reply:
You are right. Still, in this case we're passing to the caller a `nil` error . You should generate an error for this case like you are doing for the above code when you create a `NSError`.